<!DOCTYPE html>
<html>
<head>

<meta charset='utf-8'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, user-scalable=no'>
<meta name='apple-touch-fullscreen'       content='yes'>
<meta name='apple-mobile-web-app-capable' content='yes'>
<meta name='apple-mobile-web-app-status-bar-style' content='rgba(228,228,228,1.0)'>

<title>Module: ActiveJob::Exceptions::ClassMethods &mdash; Rails-master</title>

<link rel='stylesheet'  type='text/css' href='../../../css/y_fonts.css' />
<link rel='stylesheet'  type='text/css' href='../../../css/highlight.github.css' />
<link rel='stylesheet'  type='text/css' href='../../../css/y_style.css' />
<link rel='stylesheet'  type='text/css' href='../../../css/y_list.css' />
<link rel='stylesheet'  type='text/css' href='../../../css/y_color.css' />
<link rel='stylesheet'  type='text/css' href='../../../css/custom.css' />
<link rel='stylesheet'  type='text/css' href='../../../css/common.css' />

<script type='text/javascript'>
  var pathId = "ActiveJob::Exceptions::ClassMethods",
    relpath = '../../';

  var t2Info = {
    CSEP: '.',
    ISEP: '#',
    NSEP: '::'
  };
</script>

<script type='text/javascript' charset='utf-8' src='../../../js/highlight.pack.js'></script>
<script type='text/javascript' charset='utf-8' src='../../../js/y_app.js'></script>

</head>
<body>
<svg id='y_wait' class viewBox='0 0 90 90'></svg>
<div id='settings' class='hidden'></div>
<div id='y_list' class='d h'>
  <header id='list_header'></header>
  <nav id= 'list_nav' class='y_nav l_nav'>
    <ul id='list_items'></ul>
  </nav>
</div>
<div id='y_toc' class='f h'>
  <header id='toc_header'></header>
  <nav id= 'toc_nav' class='y_nav t_nav'>
  <ol id='toc_items'></ol>
  </nav>
</div>
<div id='y_main' tabindex='-1'>
  <header id='y_header'>
    <div id='y_menu'>
        <a href='../../../index.html'>Home</a> &raquo; 
        <a href='../../index.html'>Rails-master</a> &raquo; 
        <a href='../../_index.html#alpha_C'>Index (C)</a> &raquo; 
        <a href="../../ActiveJob.html" title="ActiveJob (module)">ActiveJob</a> &raquo; 
        <a href="../Exceptions.html" title="ActiveJob::Exceptions (module)">Exceptions</a> &raquo; 
        <span class='title'><a id='t2_doc_top' href='#'>ClassMethods&nbsp;&#x25B2;</a></span>
      </div>

    <a id='list_href' href="../../class_list.html"></a>
    <div id='y_measure_em' class='y_measure'></div>
    <div id='y_measure_vh' class='y_measure'></div>
    <span id='y_measure_50pre' class='y_measure'><code>123456789_123456789_123456789_123456789_123456789_</code></span>
  </header>
<div id='content' class='module'>
<h1>Module: ActiveJob::Exceptions::ClassMethods</h1>
<div id='t2_rel_sf_div'>
<table id='t2_rel_sf' class='y_box'>
  <thead>
    <tr><td id='t2_relations' colspan='2'>Relationships & Source Files</td></tr>
  </thead>
  <tbody>
    <tr><td class='box_h' colspan='2'>Extension / Inclusion / Inheritance Descendants</td></tr>
    <tr><td colspan='2'>
      <div class='box_1'>Extended In:</div>
      <div class='box_11'>
        <a class='nodoc' href="../../ActionMailer/DeliveryJob.html" title="ActionMailer::DeliveryJob (class)">::ActionMailer::DeliveryJob</a>,
          <a class='nodoc' href="../../ActionMailer/Parameterized/DeliveryJob.html" title="ActionMailer::Parameterized::DeliveryJob (class)">::ActionMailer::Parameterized::DeliveryJob</a>,
          <a href="../Base.html" title="ActiveJob::Base (class)">::ActiveJob::Base</a>
      </div>
    </td></tr>
    <tr class='single'>
      <td id='t2_defined_in' class='box_2'>Defined in:</td>
      <td class='box_rel'><a class='repo' href='https://github.com/rails/rails/blob/master/activejob/lib/active_job/exceptions.rb#L8'>activejob/lib/active_job/exceptions.rb</a>      </td>
    </tr>
  </tbody>
</table>
</div>        

<h2 class='h2_sum' id='instance_Method_summary'>Instance Method Summary</h2>
<div class='div_sum'>
<ul class='summary full'>
  <li>
    <span class='summary_signature '>
      <a href="#discard_on-instance_method" title="#discard_on (instance method)">#<strong>discard_on</strong>(exception)  </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Discard the job with no attempts to retry, if the exception is raised.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#retry_on-instance_method" title="#retry_on (instance method)">#<strong>retry_on</strong>(exception, wait: 3.seconds, attempts: 5, queue: nil, priority: nil)  </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Catch the exception and reschedule job for re-execution after so many
seconds, for a specific number of attempts.</p></div>
    </div>
  </li>
</ul>
</div>
<h2 class='y_details'>Instance Method Details</h2>
<section class='method_details first' id="discard_on-instance_method">
  <h3 class='signature  first'>
    #<strong>discard_on</strong>(exception)  
  </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Discard the job with no attempts to retry, if the exception is raised. This
is useful when the subject of the job, like an Active Record, is no longer
available, and the job is thus no longer relevant.</p>

<h4 id="label-Example">Example</h4>

<pre class="code ruby"><code class="ruby">class SearchIndexingJob &lt; {ActiveJob::Base}
  discard_on ActiveJob::DeserializationError

  def perform(record)
    # Will raise ActiveJob::DeserializationError if the record can&#39;t be deserialized
  end
end</code></pre>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/rails/rails/blob/master/activejob/lib/active_job/exceptions.rb#L71-L75'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num'>


71
72
73
74
75</pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'activejob/lib/active_job/exceptions.rb', line 71</span></pre>
<pre class='code ruby'>

<span class='kw'>def</span> <span class='id identifier rubyid_discard_on'>discard_on</span><span class='lparen'>(</span><span class='id identifier rubyid_exception'>exception</span><span class='rparen'>)</span>
  <span class='id identifier rubyid_rescue_from'>rescue_from</span> <span class='id identifier rubyid_exception'>exception</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_error'>error</span><span class='op'>|</span>
    <span class='id identifier rubyid_logger'>logger</span><span class='period'>.</span><span class='id identifier rubyid_error'>error</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Discarded </span><span class='embexpr_beg'>#{</span><span class='kw'>self</span><span class='period'>.</span><span class='id identifier rubyid_class'>class</span><span class='embexpr_end'>}</span><span class='tstring_content'> due to a </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_exception'>exception</span><span class='embexpr_end'>}</span><span class='tstring_content'>. The original exception was </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_error'>error</span><span class='period'>.</span><span class='id identifier rubyid_cause'>cause</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span><span class='embexpr_end'>}</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
  <span class='kw'>end</span>
<span class='kw'>end</span>
</pre>
  </div>
</div>
</section>

<section class='method_details' id="retry_on-instance_method">
  <h3 class='signature '>
    #<strong>retry_on</strong>(exception, wait: 3.seconds, attempts: 5, queue: nil, priority: nil)  
  </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Catch the exception and reschedule job for re-execution after so many
seconds, for a specific number of attempts. If the exception keeps getting
raised beyond the specified number of attempts, the exception is allowed to
bubble up to the underlying queuing system, which may have its own retry
mechanism or place it in a holding queue for inspection.</p>

<p>You can also pass a block that&#39;ll be invoked if the retry attempts fail
for custom logic rather than letting the exception bubble up. This block is
yielded with the job instance as the first and the error instance as the
second parameter.</p>

<h4 id="label-Options">Options</h4>
<ul><li>
<p><code>:wait</code> - Re-enqueues the job with a delay specified either in
seconds (default: 3 seconds), as a computing proc that the number of
executions so far as an argument, or as a symbol reference of
<code>:exponentially_longer</code>, which applies the wait algorithm of
<code>(executions ** 4) + 2</code> (first wait 3s, then 18s, then 83s, etc)</p>
</li><li>
<p><code>:attempts</code> - Re-enqueues the job the specified number of times
(default: 5 attempts)</p>
</li><li>
<p><code>:queue</code> - Re-enqueues the job on a different queue</p>
</li><li>
<p><code>:priority</code> - Re-enqueues the job with a different priority</p>
</li></ul>

<h4 id="label-Examples">Examples</h4>

<pre class="code ruby"><code class="ruby">class RemoteServiceJob &lt; {ActiveJob::Base}
  retry_on CustomAppException # defaults to 3s wait, 5 attempts
  retry_on AnotherCustomAppException, wait: -&gt;(executions) { executions * 2 }
  retry_on(YetAnotherCustomAppException) do |job, exception|
    ExceptionNotifier.caught(exception)
  end
  retry_on ActiveRecord::Deadlocked, wait: 5.seconds, attempts: 3
  retry_on Net::OpenTimeout, wait: :exponentially_longer, attempts: 10

  def perform(*args)
    # Might raise CustomAppException, AnotherCustomAppException, or YetAnotherCustomAppException for something domain specific
    # Might raise ActiveRecord::Deadlocked when a local db deadlock is detected
    # Might raise Net::OpenTimeout when the remote service is down
  end
end</code></pre>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/rails/rails/blob/master/activejob/lib/active_job/exceptions.rb#L43-L57'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num'>


43
44
45
46
47
48
49
50
51
52
53
54
55
56
57</pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'activejob/lib/active_job/exceptions.rb', line 43</span></pre>
<pre class='code ruby'>

<span class='kw'>def</span> <span class='id identifier rubyid_retry_on'>retry_on</span><span class='lparen'>(</span><span class='id identifier rubyid_exception'>exception</span><span class='comma'>,</span> <span class='label'>wait:</span> <span class='int'>3</span><span class='period'>.</span><span class='id identifier rubyid_seconds'>seconds</span><span class='comma'>,</span> <span class='label'>attempts:</span> <span class='int'>5</span><span class='comma'>,</span> <span class='label'>queue:</span> <span class='kw'>nil</span><span class='comma'>,</span> <span class='label'>priority:</span> <span class='kw'>nil</span><span class='rparen'>)</span>
  <span class='id identifier rubyid_rescue_from'>rescue_from</span> <span class='id identifier rubyid_exception'>exception</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_error'>error</span><span class='op'>|</span>
    <span class='kw'>if</span> <span class='id identifier rubyid_executions'>executions</span> <span class='op'>&lt;</span> <span class='id identifier rubyid_attempts'>attempts</span>
      <span class='id identifier rubyid_logger'>logger</span><span class='period'>.</span><span class='id identifier rubyid_error'>error</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Retrying </span><span class='embexpr_beg'>#{</span><span class='kw'>self</span><span class='period'>.</span><span class='id identifier rubyid_class'>class</span><span class='embexpr_end'>}</span><span class='tstring_content'> in </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_wait'>wait</span><span class='embexpr_end'>}</span><span class='tstring_content'> seconds, due to a </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_exception'>exception</span><span class='embexpr_end'>}</span><span class='tstring_content'>. The original exception was </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_error'>error</span><span class='period'>.</span><span class='id identifier rubyid_cause'>cause</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span><span class='embexpr_end'>}</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
      <span class='id identifier rubyid_retry_job'>retry_job</span> <span class='label'>wait:</span> <span class='id identifier rubyid_determine_delay'>determine_delay</span><span class='lparen'>(</span><span class='id identifier rubyid_wait'>wait</span><span class='rparen'>)</span><span class='comma'>,</span> <span class='label'>queue:</span> <span class='id identifier rubyid_queue'>queue</span><span class='comma'>,</span> <span class='label'>priority:</span> <span class='id identifier rubyid_priority'>priority</span>
    <span class='kw'>else</span>
      <span class='kw'>if</span> <span class='id identifier rubyid_block_given?'>block_given?</span>
        <span class='kw'>yield</span> <span class='kw'>self</span><span class='comma'>,</span> <span class='id identifier rubyid_exception'>exception</span>
      <span class='kw'>else</span>
        <span class='id identifier rubyid_logger'>logger</span><span class='period'>.</span><span class='id identifier rubyid_error'>error</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Stopped retrying </span><span class='embexpr_beg'>#{</span><span class='kw'>self</span><span class='period'>.</span><span class='id identifier rubyid_class'>class</span><span class='embexpr_end'>}</span><span class='tstring_content'> due to a </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_exception'>exception</span><span class='embexpr_end'>}</span><span class='tstring_content'>, which reoccurred on </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_executions'>executions</span><span class='embexpr_end'>}</span><span class='tstring_content'> attempts. The original exception was </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_error'>error</span><span class='period'>.</span><span class='id identifier rubyid_cause'>cause</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span><span class='embexpr_end'>}</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
        <span class='id identifier rubyid_raise'>raise</span> <span class='id identifier rubyid_error'>error</span>
      <span class='kw'>end</span>
    <span class='kw'>end</span>
  <span class='kw'>end</span>
<span class='kw'>end</span>
</pre>
  </div>
</div>
</section>


<div id='footer'>
  Generated by
  <a href='http://yardoc.org' title='Yay! A Ruby Documentation Tool' target="_parent">yard</a>
  0.9.8 with <a href='https://msp-greg.github.io/yard-t2/' title='yard-t2' target="_parent">yard-t2</a> 0.9.0 (ruby-2.5.0p-1)
</div>
</div> <!-- content  -->
</div> <!-- y_main   -->
</body>
</html>